File: exercises/day15-No-Exceptions-For-Failures/typescript/src/Business.test.ts
// Business.test.ts
import { Business, BusinessError } from "./Business";
import { Child, Gift } from "./models";
import { Factory, Inventory, WishList } from "./dependencies";

function loadGiftAndExtractInnerErrorMessage(business: Business, child: Child): string {
  try {
    business.loadGiftsInSleigh(child);
  } catch (e) {
    if (e instanceof BusinessError && e.cause) {
      return e.cause.message;
    }
    throw e;
  }
  throw new Error("Expected BusinessError to be thrown");
}

test("child wish not found error message is extracted", () => {
  const timmy = new Child("Timmy");

  const wishList: WishList = {
    identifyGift: () => null,
  };
  const factory: Factory = {
    findManufacturedGift: () => null,
  };
  const inventory: Inventory = {
    pickUpGift: () => null,
  };

  const business = new Business(factory, inventory, wishList);

  const message = loadGiftAndExtractInnerErrorMessage(business, timmy);

  expect(message).toBe("No wish found for child: Timmy");
});

test("gift not manufactured error message is extracted", () => {
  const timmy = new Child("Timmy");
  const wishedGift = new Gift("Lego Death Star", "BARCODE-123");

  const wishList: WishList = {
    identifyGift: () => wishedGift,
  };
  const factory: Factory = {
    findManufacturedGift: () => null,
  };
  const inventory: Inventory = {
    pickUpGift: () => null,
  };

  const business = new Business(factory, inventory, wishList);

  const message = loadGiftAndExtractInnerErrorMessage(business, timmy);

  expect(message).toBe("Gift has not been manufactured: Lego Death Star");
});

test("gift out of stock error message is extracted", () => {
  const timmy = new Child("Timmy");
  const wishedGift = new Gift("Red Bike", "BARCODE-456");
  const manufacturedGift = new Gift("Red Bike", "BARCODE-456");

  const wishList: WishList = {
    identifyGift: () => wishedGift,
  };
  const factory: Factory = {
    findManufacturedGift: () => manufacturedGift,
  };
  const inventory: Inventory = {
    pickUpGift: () => null,
  };

  const business = new Business(factory, inventory, wishList);

  const message = loadGiftAndExtractInnerErrorMessage(business, timmy);

  expect(message).toBe("Gift out of stock: Red Bike");
});

File: exercises/day15-No-Exceptions-For-Failures/scala/src/test/scala/santaChristmasList/operations/BusinessTest.scala
package santaChristmasList.operations

import org.scalatest.funsuite.AnyFunSuite
import santaChristmasList.operations.dependencies._
import santaChristmasList.operations.models._

class BusinessTest extends AnyFunSuite {

  test("child wish not found error message is extracted") {
    val timmy = Child("Timmy")

    val wishList = new WishList {
      override def identifyGift(child: Child): Gift = null
    }
    val factory = new Factory {
      override def findManufacturedGift(gift: Gift): Gift = null
    }
    val inventory = new Inventory {
      override def pickUpGift(barCode: String): Gift = null
    }

    val business = new Business(factory, inventory, wishList)

    val msg = loadGiftAndExtractInnerErrorMessage(business, timmy)

    assert(msg == "No wish found for child: Timmy")
  }

  test("gift not manufactured error message is extracted") {
    val timmy = Child("Timmy")
    val wishedGift = Gift("Lego Death Star", "BARCODE-123")

    val wishList = new WishList {
      override def identifyGift(child: Child): Gift = wishedGift
    }
    val factory = new Factory {
      override def findManufacturedGift(gift: Gift): Gift = null
    }
    val inventory = new Inventory {
      override def pickUpGift(barCode: String): Gift = null
    }

    val business = new Business(factory, inventory, wishList)

    val msg = loadGiftAndExtractInnerErrorMessage(business, timmy)

    assert(msg == "Gift has not been manufactured: Lego Death Star")
  }

  test("gift out of stock error message is extracted") {
    val timmy = Child("Timmy")
    val wishedGift = Gift("Red Bike", "BARCODE-456")
    val manufacturedGift = Gift("Red Bike", "BARCODE-456")

    val wishList = new WishList {
      override def identifyGift(child: Child): Gift = wishedGift
    }
    val factory = new Factory {
      override def findManufacturedGift(gift: Gift): Gift = manufacturedGift
    }
    val inventory = new Inventory {
      override def pickUpGift(barCode: String): Gift = null
    }

    val business = new Business(factory, inventory, wishList)

    val msg = loadGiftAndExtractInnerErrorMessage(business, timmy)

    assert(msg == "Gift out of stock: Red Bike")
  }

  private def loadGiftAndExtractInnerErrorMessage(business: Business, child: Child): String = {
    val ex = intercept[BusinessException] {
      business.loadGiftsInSleigh(child)
    }
    val cause = ex.getCause
    assert(cause != null)
    cause.getMessage
  }
}
